.github/workflows/pypi-package.yml

```diff
@@ -0,0 +1,67 @@
+---
+name: Build & maybe upload PyPI package
+
+on:
+  push:
+    branches: [main]
+    tags: ["*"]
+  pull_request:
+    branches: [main]
+  release:
+    types:
+      - published
+  workflow_dispatch:
+
+permissions:
+  contents: read
+  id-token: write
+
+jobs:
+  # Always build & lint package.
+  build-package:
+    name: Build & verify package
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - uses: hynek/build-and-inspect-python-package@v1
+
+  # Upload to Test PyPI on every commit on main.
+  release-test-pypi:
+    name: Publish in-dev package to test.pypi.org
+    if: github.repository_owner == 'python' && github.event_name == 'push' && github.ref == 'refs/heads/main'
+    runs-on: ubuntu-latest
+    needs: build-package
+
+    steps:
+      - name: Download packages built by build-and-inspect-python-package
+        uses: actions/download-artifact@v3
+        with:
+          name: Packages
+          path: dist
+
+      - name: Upload package to Test PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+
+  # Upload to real PyPI on GitHub Releases.
+  release-pypi:
+    name: Publish released package to pypi.org
+    environment: release-pypi
+    if: github.repository_owner == 'python' && github.event.action == 'published'
```

Comment:
Yep.

We must always create a tag for releases.

We should also create a GitHub Release, because people can subscribe to those notifications in GitHub, but not tags. 

I will later update the release instructions.

Reply:
Sorry, my poor phrasing -- which bit are you saying yes to?

As I understand it, we want the 'release' action to be ``git tag vYYY.MM.DD -m "..."; git push upstream vYYY.MM.DD``, but I don't believe that this would satisgy the ``github.event.action == 'published'`` condition?

We could use e.g. actions/github-script, softprops/action-gh-release, ncipollo/release-action to create a release via GHA when a tag is pushed?

Reply:
Ah sorry, I wasn't very clear either :)

I would suggest the release process to be:

1. Manually update version numbers as before
2. Go to https://github.com/python/python-docs-theme/releases
3. Click "Draft a new release"
4. Click "Choose a tag", type the next version (for example "2023.08") and select "Create new tag: 2023.8 on publish"
5. Click "Generate release notes" and amend as required
6. Click "Publish release"

This then creates the Git tag and the GitHub Release all at once, and triggers GitHub Actions to upload to PyPI.

Reply:
Would creating the tag via GH releases create a 'proper' annotated tag (i.e. what ``git tag -m`` would create) or just a lightweight tag?

Reply:
I'm not sure the difference :)

It will create ones like in:

* https://github.com/jazzband/prettytable/tags
* https://github.com/jazzband/prettytable/releases

What kind are they? What kind do we have in this repo?

<details>
<summary>Automation aside</summary>

That repo has more automation: 

I add labels like "Changelog: Added", "Changelog: Fixed", "Changelog: skip" to PRs.

These are then used as headings in the GitHub release.

Based on the labels, the next SemVer version is chosen automatically.

I don't need to do anything from the command line to release, just publish an already drafted release, with some light editing if necessary. From the repo homepage, three clicks and it's released. Which means I can release from any browser including my phone!
</details>

Reply:
Annotated tags are full objects in git, whereas lightweight tags are just pointers to commits [^1]. From the documentation of ``git tag`` [^2]: (emphasis mine)

> **Annotated tags are meant for release** while lightweight tags are meant for private or temporary object labels. For this reason, some git commands for naming objects (like ``git describe``) will ignore lightweight tags by default.

> What kind are they?

Running ``git for-each-ref refs/tags`` shows the difference:

```console
/tmp/prettytable$ git for-each-ref refs/tags
32d4cd30f84fba75ddb1b1bcb2f824acd915146a commit refs/tags/0.7.0
af9737e055e69ef70d25172eca375988080690e5 commit refs/tags/0.7.1
7be1937eebe75775d429f10661cde04c3509a9aa commit refs/tags/0.7.2
464c5306f6c1c982de52523a8baa464a60434c00 tag    refs/tags/1.0.0
6563a29c32c2b3e4cc8bc924197151d723559caf commit refs/tags/1.0.1
9a7c7484e0b6b932f7122f762d2257fad2a18f3a commit refs/tags/2.0.0
bc7e5319992304d820249e0530050671f217ef1d commit refs/tags/2.1.0
1147cc4e02c244fca26057c9401fff839eb94cc3 commit refs/tags/2.2.0
5ce1d24d764aaff936077d0212bad56bd9626972 commit refs/tags/2.2.1
[...snip]
```

``tag`` is an annotated tag, whereas ``commit`` is a lightweight tag.

Running a similar command to show tag / commit messages ("a" in the left-hand column means "annotated tag"):

```console
/tmp/prettytable$ git for-each-ref --format="%(if:equals=tag)%(objecttype)%(then)a %(else)  %(end)%(align:10,right)%(refname:short)%09%(objectname:short)%(end)%09%(contents:subject)" refs/tags
  0.7.0 32d4cd3 Copied 0.7-RELEASE code from 0.7 branch to trunk.\n\nDon't count ANSI color codes when calculating string length.
  0.7.1 af9737e Added xhtml option to control linebreak element produced by get_html_string().
  0.7.2 7be1937 Added print_empty option to control whether or not emtpy tables result in a blank string or an empty frame.
a 1.0.0 464c530 Release 1.0.0
  1.0.1 6563a29 See GitHub Releases for 1.0.1 and newer
  2.0.0 9a7c748 Remove Travis CI from release checklist
  2.1.0 bc7e531 Merge pull request #96 from hugovk/importlib-version
  2.2.0 1147cc4 Merge pull request #125 from hugovk/add-3.10
  2.2.1 5ce1d24 Merge pull request #132 from hydrargyrum/master
[..snip]
```

Running the same command in e.g. CPython gives a list of "properly" named and tagged commits (though sorted wrong):

```
[...snip]
a v3.11.0       968953dcd89     Python 3.11.0
a v3.11.0a1     a510de09534     Python 3.11.0a1
a v3.11.0a2     ae7760f2697     Python 3.11.0a2
a v3.11.0a3     b948add7910     Python 3.11.0a3
a v3.11.0a4     95b215f5b22     Python 3.11.0a4
a v3.11.0a5     838b51a0799     Python 3.11.0a5
a v3.11.0a6     5191ba090ef     Python 3.11.0a6
a v3.11.0a7     45772541f67     Python 3.11.0a7
a v3.11.0b1     34ecdf95f32     Python 3.11.0b1
a v3.11.0b2     e2be1865f95     Python 3.11.0b2
a v3.11.0b3     70dfc75cef6     Python 3.11.0b3
a v3.11.0b4     c756c17b892     Python 3.11.0b4
a v3.11.0b5     640c7832151     Python 3.11.0b5
a v3.11.0rc1    51db56ccbb1     Python 3.11.0rc1
a v3.11.0rc2    484c8183928     Python 3.11.0rc2
a v3.11.1       92f76e215d5     Python 3.11.1
a v3.11.2       814da78353b     Python 3.11.2
a v3.11.3       edb4401c7fc     Python 3.11.3
a v3.11.4       c92d5e040a3     Python 3.11.4
a v3.11.5       7b52b76d138     Python 3.11.5
[...snip]
```

> What kind do we have in this repo?

This repo currently has only lightweight tags, but my ambition would be to use "proper" release tags instead -- so using ``git tag -m`` instead of plain ``git tag``.

See also this unresolved [feature request](https://github.com/orgs/community/discussions/4924) for GitHub releases to create annotated tags.

[^1]: https://git-scm.com/book/en/v2/Git-Basics-Tagging#_creating_tags
[^2]: https://git-scm.com/docs/git-tag#_description

Reply:
Oh, thanks, but I'm not worried about that: we've already been using the "wrong" kind of tags in this repo with no problem, and I have in other projects, and GitHub has chosen it for Releases which will have been used who knows how many hundreds of thousands of times for who knows how many thousands of projects.

I'd instead suggest asking GitHub to change the default so it can benefit all users.

Reply:
Referencing https://github.com/pypa/setuptools_scm/issues/521, given you referenced ``setuptools_scm`` -- it seems if we do e.g. a ``rc1`` release and then a ``.0`` release, where both tags reference the same commit, tools like ``setuptools_scm`` may fail. Perhaps unlikely for this repo given date-based releases but something to be aware of.

Reply:
That's not a concern.

We're not using `setuptools_scm` or `hatch-vcs`, and we've never made an RC in this repo.

But if:
* we do start using `setuptools_scm` or `hatch-vcs`,
* _and_ then want to make an RC, 
* _and_ then want to make a `.0` release without other changes, 
* _and_ it's with a potentially clashing CalVer number,

Then we'll have to either:
* delete the redundant RC tag first, 
* or create a trivial commit first, 
* or do a `.1` or something.

But that situation is several hypotheticals and facts away from where we are, and it's easily solvable.

Reply:
True. I suppose I'd like to see us doing the 'right' thing w.r.t. Git, but that GitHub by default uses the discouraged form is unhelpful.

As you point out, though, what GitHub does is the status quo for this repo, so don't consider this a blocking concern.

A